src/sys/inotify.rs

```diff
@@ -137,6 +117,43 @@ bitflags! {
     }
 }
 
+#[derive(Debug)]
+pub struct InotifyEvent {
+    pub wd: i32,
+    pub mask: EventFlags,
+    pub cookie: u32,
+    pub name: String
+}
+
+pub fn inotify_read_event(fd: RawFd) -> Result<InotifyEvent> {
+    let mut buffer = [0u8; 4096];
+
+    read(fd, &mut buffer)?;
+
+    let event_size = std::mem::size_of::<ffi::inotify_event>();
+    assert!(buffer.len() >= event_size);
+
+    let event = buffer.as_ptr() as *const ffi::inotify_event;
+    let event = unsafe { &*event };
+
+    let bytes_left_in_buffer = buffer.len() - event_size;
```

Comment:
Instead of doing this trick, you could just make the FFI `inotify_event` struct's last field be an unsized slice.  That's what the Linux headers do.  Like this:
```rust
struct inotify_event<T: ?Sized> {
    pub stuff: Stuff,
    pub name: T
}
let header_size = mem::size_of::<inotify_event<[u8; 0]>>();
let name_size = buffer.len() - header_size;
let event = buffer.as_ptr() as *const inotify_event<[u8; name_size]>()
```

Reply:
I'm not sure I understand your example correctly, `name_size` is not constant therefore `[u8; name_size]` is invalid, no ?

I do understand the general meaning though.

Reply:
It would be invalid to construct a `[u8; name_size]` and store it on the stack.  However, a pointer to an array of unknown size *does* have a constant size; it's a pointer.  So the expression is valid.

Reply:
Ok I think I am missing something because this doesn't compile. 

I'm sorry if my level in Rust is not great; I'm coming from C.

I made a small rust playground with the smallest code necessary to reproduce:

[](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=45520d14233c767d2fcb0b22b0e730b0)

Reply:
Well, I guess I was wrong.  In that case, you'll probably have to use slice::from_raw_parts.

Reply:
So, I refactored some of that code but I still end up with a problem. The name_size is actually aligned which means that my string ends up with a lot of trailing NULLs in some cases. The original code for trimming it was ugly : 
```
let name = name.splitn(2, |b| b == &0u8).next().unwrap();
```
However I can't find a much better way. `OsStr::from_bytes` doesn't work here.

I found this thread on StackOverflow, but it is arguably uglier: [https://stackoverflow.com/questions/42066381/how-to-get-a-str-from-a-nul-terminated-byte-slice-if-the-nul-terminator-isnt](https://stackoverflow.com/questions/42066381/how-to-get-a-str-from-a-nul-terminated-byte-slice-if-the-nul-terminator-isnt)

Reply:
What's wrong with `OsStrExt::from_bytes`?

Reply:
`OsStrExt::from_bytes` does not take of NULL bytes. It happily accepts to have NULL bytes in the middle of the resulting `OsStr`. In my case, I end up with trailing NULLs.

Reply:
Oh, I forgot that it requires you to take care of that yourself.  Well, there's always `CStr::from_ptr`.

Reply:
Ok fixed it with CStr::from_ptr.